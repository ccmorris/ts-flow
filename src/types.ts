export type ActivityPointer = string
export type EndPointer = null

export type CatchDefinition = {
  [key: string]: {
    then: ActivityPointer | EndPointer
  }
}

export type ActivityContext = {
  fail: (error: string) => void
}

export type ActivityFunction = (
  input: unknown,
  context: ActivityContext
) => Promise<unknown> // TODO: more type safe input/output

export type ActivityDefinition = {
  name: string
  start?: true
  fn: ActivityFunction
  then: ActivityPointer | EndPointer
  catch?: CatchDefinition
}

export type ActivityDefinitions = ActivityDefinition[]
