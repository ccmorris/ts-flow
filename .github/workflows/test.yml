# Run unit tests with Bun on every push

name: CI Tests

on:
  push:
    branches:
      - '*'
  pull_request:
    branches:
      - main

jobs:
  test:
    runs-on: ubuntu-latest

    steps:
      - name: Checkout repository
        uses: actions/checkout@v3

      - name: Set up Bun
        uses: oven-sh/setup-bun@v1
        with:
          bun-version: latest # You can specify a version like '1.0.0' if needed

      - name: Install dependencies
        run: bun install

      - name: Run tests
        run: bun test
