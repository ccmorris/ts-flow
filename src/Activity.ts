import type {
  TaskDefinitions,
  ActivityFunction,
  EndPointer,
  CatchInput,
  ActivityContext,
} from './types'
import type { Choice } from './Choice'

/**
 * An activity is a single unit of work in a workflow.
 * Define an async function, and chain activities together to create a workflow.
 */
export class Activity<I, O, C extends ActivityContext> {
  name: string
  fn: ActivityFunction<I, O, C>
  next?: Activity<O, unknown, C> | Choice<O, unknown, C> | EndPointer
  catchConfig?: {
    [key: string]: {
      then:
        | Activity<CatchInput<any>, unknown, C>
        | Choice<CatchInput<any>, unknown, C>
        | EndPointer
    }
  }

  public constructor(name: string, fn: ActivityFunction<I, O, C>) {
    this.name = name
    this.fn = fn
  }

  /**
   * Chain another activity or choice to run after this one, when successful.
   */
  public then<T extends Activity<O, any, C> | Choice<O, any, C>>(next: T): T {
    this.next = next
    return next
  }

  /**
   * Chain a catch activity or choice to run after this one, when an error occurs.
   *
   * @example Catch an ErrorName and run an activity:
   * activity.catch('ErrorName', new Activity('activity', async () => {}))
   *
   * @example Catch an ErrorName and end the workflow as a success:
   * activity.catch('ErrorName', null)
   */
  public catch<CatchKey extends string>(
    error: CatchKey,
    next:
      | Activity<CatchInput<CatchKey>, any, C>
      | Choice<CatchInput<CatchKey>, any, C>
      | EndPointer
  ) {
    this.catchConfig = { ...this.catchConfig, [error]: { then: next } }
    return this
  }

  /**
   * Convert the activity and any chained activities to a list of task definitions.
   */
  public toTaskDefinitions(): TaskDefinitions {
    const tasks = [
      {
        type: 'activity' as const,
        name: this.name,
        fn: this.fn,
        then: this.next?.name ?? null,
        catch: this.catchConfig
          ? Object.fromEntries(
              Object.entries(this.catchConfig).map(([error, catchTask]) => {
                return [error, { then: catchTask.then?.name ?? null }]
              })
            )
          : undefined,
      },
      ...(this.next ? this.next.toTaskDefinitions() : []),
      ...(this.catchConfig
        ? Object.values(this.catchConfig)
            .map((catchTask) => catchTask.then?.toTaskDefinitions())
            .filter((task) => !!task)
            .flat()
        : []),
    ]
    return removeDuplicateTasks(tasks as TaskDefinitions)
  }
}

// removes duplicates from an array using the name property
const removeDuplicateTasks = (tasks: TaskDefinitions) => {
  const seen = new Set()
  return tasks.filter((task) => {
    if (seen.has(task.name)) return false
    seen.add(task.name)
    return true
  })
}
