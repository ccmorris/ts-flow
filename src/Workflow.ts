import type { Activity } from './Activity'
import type { Choice } from './Choice'
import { toDiagramPngUrl, toMermaid } from './diagrams'
import { run } from './run'
import type { WorkflowResult } from './types'

export type Task<I, O> = Activity<I, O> | Choice<I, O>

export class Workflow<I> {
  startTask: Task<I, any>

  constructor({ startTask }: { startTask: Task<I, any> }) {
    this.startTask = startTask
  }

  public async run(initialInput: I): Promise<WorkflowResult> {
    return run(this.startTask.toTaskDefinitions(), initialInput)
  }

  public toMermaid(): string {
    return toMermaid(this.startTask.toTaskDefinitions())
  }
  public toPngUrl(workflowResult?: WorkflowResult): string {
    return toDiagramPngUrl(this.startTask.toTaskDefinitions(), workflowResult)
  }
}
