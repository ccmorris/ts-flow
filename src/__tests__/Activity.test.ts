import { describe, expect, mock, test } from 'bun:test'

import { Activity } from '../Activity'
import { CatchInput } from '../types'
import { Workflow } from '../Workflow'

describe('Activity class', () => {
  test('should create an activity', () => {
    const activity = new Activity('activity', async () => {})

    expect(activity.name).toBe('activity')
    expect(activity.fn).toBeInstanceOf(Function)
  })

  test('should chain activities', () => {
    const activity1 = new Activity('activity1', async (_: '1') => '2' as const)
    const activity2 = new Activity('activity2', async (_: '2') => '3' as const)
    const activity3 = new Activity('activity3', async (_: '3') => '4' as const)

    activity1.then(activity2).then(activity3)

    expect(activity1.next?.name).toBe(activity2.name)
    expect(activity2.next?.name).toBe(activity3.name)
  })

  test('should chain activities with catch', () => {
    const activity1 = new Activity('activity1', async (_: '1') => '2' as const)
    const activity2 = new Activity('activity2', async (_: '2') => '3' as const)
    const activity3 = new Activity(
      'activity3',
      async (_: CatchInput<'error'>) => '4' as const
    )

    activity1.then(activity2).catch('error', activity3)

    expect(activity1.next?.name).toBe(activity2.name)
    expect(activity2.catchConfig).toEqual({ error: { then: activity3 as any } })
  })

  test('should convert to activity definitions', () => {
    const activity1 = new Activity('activity1', async () => {})
    const activity2 = new Activity('activity2', async () => {})
    const activity3 = new Activity('activity3', async () => {})

    activity1.then(activity2).catch('error', activity3)

    const taskDefinitions = activity1.toTaskDefinitions()

    expect(taskDefinitions).toEqual([
      {
        type: 'activity',
        name: 'activity1',
        fn: activity1.fn,
        then: 'activity2',
      },
      {
        type: 'activity',
        name: 'activity2',
        fn: activity2.fn,
        then: null,
        catch: { error: { then: 'activity3' } },
      },
      {
        type: 'activity',
        name: 'activity3',
        fn: activity3.fn,
        then: null,
      },
    ])
  })

  test('should run activities', async () => {
    const activity1 = new Activity(
      'activity1',
      mock().mockResolvedValue('output from activity1')
    )
    const activity2 = new Activity(
      'activity2',
      mock().mockRejectedValue(new Error('Timeout'))
    )
    const activity3 = new Activity(
      'activity3',
      mock().mockResolvedValue('output from activity3')
    )

    activity1.then(activity2).catch('Timeout', activity3)

    const result = await new Workflow({ startTask: activity1 }).run(
      'initial input'
    )

    expect(activity1.fn).toHaveBeenCalledTimes(1)
    expect(activity1.fn).toHaveBeenCalledWith(
      'initial input',
      expect.any(Object)
    )
    expect(activity2.fn).toHaveBeenCalledTimes(1)
    expect(activity2.fn).toHaveBeenCalledWith(
      'output from activity1',
      expect.any(Object)
    )
    expect(activity3.fn).toHaveBeenCalledTimes(1)
    expect(activity3.fn).toHaveBeenCalledWith(
      expect.objectContaining({ key: 'Timeout', error: expect.any(Error) }),
      expect.any(Object)
    )

    expect(result).toMatchSnapshot()
  })
})
