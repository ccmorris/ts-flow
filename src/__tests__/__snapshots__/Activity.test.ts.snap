// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Activity class should run activities 1`] = `
{
  "context": {},
  "output": "output from activity3",
  "success": true,
  "transitions": [
    {
      "from": null,
      "nextInput": "initial input",
      "to": {
        "catch": undefined,
        "fn": [class Function],
        "name": "activity1",
        "then": "activity2",
        "type": "activity",
      },
      "transitionName": "(start)",
    },
    {
      "from": {
        "catch": undefined,
        "fn": [class Function],
        "name": "activity1",
        "then": "activity2",
        "type": "activity",
      },
      "nextInput": "output from activity1",
      "to": {
        "catch": {
          "Timeout": {
            "then": "activity3",
          },
        },
        "fn": [class Function],
        "name": "activity2",
        "then": null,
        "type": "activity",
      },
      "transitionName": "then",
    },
    {
      "from": {
        "catch": {
          "Timeout": {
            "then": "activity3",
          },
        },
        "fn": [class Function],
        "name": "activity2",
        "then": null,
        "type": "activity",
      },
      "nextInput": {
        "error": [Error: Timeout],
        "key": "Timeout",
      },
      "to": {
        "catch": undefined,
        "fn": [class Function],
        "name": "activity3",
        "then": null,
        "type": "activity",
      },
      "transitionName": "Timeout",
    },
    {
      "from": {
        "catch": undefined,
        "fn": [class Function],
        "name": "activity3",
        "then": null,
        "type": "activity",
      },
      "nextInput": "output from activity3",
      "to": null,
      "transitionName": "(end)",
    },
  ],
}
`;
